src/support/bitset.h

```diff
@@ -0,0 +1,216 @@
+// std::set or std::unordered_set as closely as possible.
+class BitSet : std::vector<bool> {
+  std::vector<bool>& vec() noexcept { return *this; }
+  const std::vector<bool>& vec() const noexcept { return *this; }
+
+  void ensure(size_t value) noexcept {
+    if (vec().size() <= value) {
+      vec().resize(value + 1);
+    }
+  }
+
+public:
+  using key_type = size_t;
+  using value_type = size_t;
+
+  class iterator {
+    const std::vector<bool>* parent = nullptr;
+    size_t index = 0;
+
+    iterator(const std::vector<bool>* parent) : parent(parent) {
+      const auto size = parent->size();
+      while (index < size && !(*parent)[index]) {
+        ++index;
+      }
+    }
+
+    iterator(const std::vector<bool>* parent, size_t index)
+      : parent(parent), index(index) {}
+
+    friend BitSet;
+
+  public:
+    using difference_type = ptrdiff_t;
+    using value_type = size_t;
+    using pointer = size_t*;
+    using reference = size_t&;
+    using iterator_category = std::bidirectional_iterator_tag;
+
+    iterator() = default;
+    iterator(const iterator&) = default;
+    iterator(iterator&&) = default;
+    iterator& operator=(const iterator&) = default;
+    iterator& operator=(iterator&&) = default;
+
+    size_t operator*() const noexcept { return index; }
+
+    iterator& operator++() noexcept {
+      const auto size = parent->size();
+      do {
+        ++index;
+      } while (index < size && !(*parent)[index]);
+      return *this;
+    }
+
+    iterator operator++(int) noexcept {
+      iterator it = *this;
+      ++(*this);
+      return it;
+    }
+
+    iterator& operator--() noexcept {
+      do {
+        --index;
+        assert(index != -1ull);
+      } while (!(*parent)[index]);
+      return *this;
+    }
+
+    iterator operator--(int) noexcept {
+      iterator it = *this;
+      --(*this);
+      return it;
+    }
+
+    bool operator==(const iterator& other) const noexcept {
+      assert(parent == other.parent);
+      return index == other.index;
+    }
+    bool operator!=(const iterator& other) const noexcept {
+      return !(*this == other);
+      assert(parent == other.parent);
+      return index == other.index;
+    }
+  };
+
+  BitSet() = default;
+  BitSet(const BitSet&) = default;
+  BitSet(BitSet&&) = default;
+
+  template<typename It> BitSet(It begin, It end) {
+    for (auto it = begin; it != end; ++it) {
+      insert(*it);
+    }
+  }
+  BitSet(const std::initializer_list<size_t>& vals)
+    : BitSet(vals.begin(), vals.end()) {}
+
+  BitSet& operator=(const BitSet&) = default;
+  BitSet& operator=(BitSet&&) = default;
+
+  iterator begin() const noexcept { return iterator({this}); }
+  iterator end() const noexcept { return iterator({this, vec().size()}); }
+
+  bool empty() const noexcept { return begin() == end(); }
+
+  size_t size() const noexcept {
+    size_t result = 0;
+    for (auto it = begin(); it != end(); ++it, ++result) {
+    }
```

Comment:
Wouldn't it be possibly simpler and faster to iterate on the underlying `std::vector<bool>`?
```cpp
auto* super = (std::vector<bool>*)this;
for (auto x : super) result += x;

Reply:
(possibly simpler & faster since the current code will have nested loops)

Reply:
I would hope it would optimize to the same thing, but yes, I can rewrite the code to use the underlying vector directly.